Handle cases where the output payload type is different for the input content-type

For example the following pipeline that uses `file` source to read file’s content (octet type) applies `transform` to extract the file names form the header, which practically alters the content type to text:
```
file --filename-pattern='*.log' --directory=/tmp --prevent-duplicates=false | transform --expression=headers[file_name] | log
```
This fails in the transform module:
```
Caused by: java.lang.IllegalStateException: Failed to convert message: 'GenericMessage [payload=com.wacom.TabletHelper.so.log, headers={amqp_receivedDeliveryMode=PERSISTENT, amqp_receivedExchange=filetest.file, amqp_deliveryTag=2, file_name=com.wacom.TabletHelper.so.log, deliveryAttempt=3, amqp_consumerQueue=filetest.file.filetest, amqp_redelivered=false, file_originalFile=/tmp/com.wacom.TabletHelper.so.log, file_relativePath=com.wacom.TabletHelper.so.log, amqp_receivedRoutingKey=filetest.file, id=789348da-4bde-d242-a772-6bec0184b939, amqp_consumerTag=amq.ctag-acFRbva_SKTQqJsSv_Eywg, contentType=application/octet-stream, timestamp=1526295823965}]' to outbound message.
        at org.springframework.cloud.stream.binding.MessageConverterConfigurer$OutboundContentTypeConvertingInterceptor.doPreSend(MessageConverterConfigurer.java:324) ~[spring-cloud-stream-2.0.0.RELEASE.jar!/:2.0.0.RELEASE]
        at org.springframework.cloud.stream.binding.MessageConverterConfigurer$AbstractContentTypeInterceptor.preSend(MessageConverterConfigurer.java:351) ~[spring-cloud-stream-2.0.0.RELEASE.jar!/:2.0.0.RELEASE]
        at org.springframework.integration.channel.AbstractMessageChannel$ChannelInterceptorList.preSend(AbstractMessageChannel.java:589) ~[spring-integration-core-5.0.4.RELEASE.jar!/:5.0.4.RELEASE]
        at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:435) ~[spring-integration-core-5.0.4.RELEASE.jar!/:5.0.4.RELEASE]
```
( the `com.wacom.TabletHelper.so.log` is the file name) 
Exception comes from https://goo.gl/mgaXtd  and is likely related to the fact that the message payload has changed to `text` but the `content-type` in the header is still `octet`.  Later combination doesn’t match with any of the pre-configured message converters.
The solution used for the Log app will not help here! Instead to solve this we have to change the output content type (either explicitly with new property or find some smart way to determine the right output type).

--- a/spring-cloud-starter-stream-processor-transform/src/main/java/org/springframework/cloud/stream/app/transform/processor/TransformProcessorConfiguration.java
+++ b/spring-cloud-starter-stream-processor-transform/src/main/java/org/springframework/cloud/stream/app/transform/processor/TransformProcessorConfiguration.java
@@ -20,7 +20,7 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.cloud.stream.annotation.EnableBinding;
 import org.springframework.cloud.stream.messaging.Processor;
-import org.springframework.integration.annotation.Transformer;
+import org.springframework.integration.annotation.ServiceActivator;
 import org.springframework.messaging.Message;
 
 /**
@@ -37,7 +37,7 @@ public class TransformProcessorConfiguration {
 	@Autowired
 	private TransformProcessorProperties properties;
 
-	@Transformer(inputChannel = Processor.INPUT, outputChannel = Processor.OUTPUT)
+	@ServiceActivator(inputChannel = Processor.INPUT, outputChannel = Processor.OUTPUT)
 	public Object transform(Message<?> message) {
 		return properties.getExpression().getValue(message);
 	}

--- a/spring-cloud-starter-stream-processor-transform/src/test/java/org/springframework/cloud/stream/app/transform/processor/TransformProcessorIntegrationTests.java
+++ b/spring-cloud-starter-stream-processor-transform/src/test/java/org/springframework/cloud/stream/app/transform/processor/TransformProcessorIntegrationTests.java
@@ -16,6 +16,9 @@
 
 package org.springframework.cloud.stream.app.transform.processor;
 
+import java.util.HashMap;
+import java.util.Map;
+
 import org.junit.Test;
 import org.junit.runner.RunWith;
 
@@ -26,6 +29,7 @@ import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.cloud.stream.messaging.Processor;
 import org.springframework.cloud.stream.test.binder.MessageCollector;
 import org.springframework.integration.config.SpelFunctionFactoryBean;
+import org.springframework.messaging.MessageHeaders;
 import org.springframework.messaging.support.GenericMessage;
 import org.springframework.test.annotation.DirtiesContext;
 import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
@@ -83,6 +87,19 @@ public abstract class TransformProcessorIntegrationTests {
 			assertNotNull(this.xpathSpelFunctionFactoryBean);
 			this.channels.input().send(new GenericMessage<Object>("hello"));
 			assertThat(this.collector.forChannel(this.channels.output()), receivesPayloadThat(is("HELLO")));
+		}
+	}
+
+	@SpringBootTest("transformer.expression=new String(payload)")
+	public static class PayloadTypeDifferentFromOutputContentTypeTests extends TransformProcessorIntegrationTests {
+
+		@Test
+		public void test() {
+			Map<String, Object> headers = new HashMap<>();
+			headers.put(MessageHeaders.CONTENT_TYPE, "application/octet-stream");
+			channels.input().send(new GenericMessage<Object>("hello".getBytes(), new MessageHeaders(headers)));
+
+			assertThat(collector.forChannel(channels.output()), receivesPayloadThat(is("hello")));
 		}
 	}
 

